torch/csrc/distributed/rpc/py_rref.cpp

```diff
@@ -0,0 +1,107 @@
+#include <torch/csrc/distributed/rpc/py_rref.h>
+
+#include <torch/csrc/distributed/rpc/rref_context.h>
+#include <torch/csrc/jit/pybind_utils.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+namespace {
+
+// Constants below are used in PyRRef pickling and unpickling. PyRRef is
+// converted into a py::tuple in pickling, and reconstructed from the py::tuple
+// in unpickling.
+constexpr int RREF_TUPLE_SIZE = 2; // number of data fields in the py::tuple
+constexpr int RFD_IDX = 0; // index of RRefForkData
+constexpr int TYPE_IDX = 1; // index of type (py::object or IValue)
+
+} // namespace
+
+PyRRef::PyRRef(std::shared_ptr<RRef> rref) : rref_(std::move(rref)) {
+  TORCH_CHECK(rref_, "PyRRef must not wrap nullptr");
+}
+
+bool PyRRef::isOwner() const {
+  return rref_->isOwner();
+}
+
+worker_id_t PyRRef::owner() const {
+  return rref_->owner();
+}
+
+py::object PyRRef::toHere() {
+  if (rref_->isPyObj()) {
+    if (rref_->isOwner()) {
+      const py::object& value =
+          std::static_pointer_cast<OwnerRRef<py::object>>(rref_)->getValue();
+      // acquiring GIL as the return statement construct a new py::object from
+      // a const reference.
+      AutoGIL ag;
+      return value;
+    } else {
+      return std::static_pointer_cast<UserRRef<py::object>>(rref_)->toHere();
+    }
+  } else {
+    IValue value;
+    if (rref_->isOwner()) {
+      value = std::static_pointer_cast<OwnerRRef<IValue>>(rref_)->getValue();
+    } else {
+      value = std::static_pointer_cast<UserRRef<IValue>>(rref_)->toHere();
+    }
+    // acquiring GIL as torch::jit::toPyObject creates new py::object without
+    // grabbing the GIL.
+    AutoGIL ag;
+    return torch::jit::toPyObject(std::move(value));
+  }
+}
+
+py::object PyRRef::localValue() {
+  TORCH_CHECK(
+      rref_->isOwner(),
+      "Cannot call localValue() on a non-local reference. Call it on ",
+      RRefContext::getInstance()->getWorkerName());
+
+  if (rref_->isPyObj()) {
+    const py::object& value =
+        std::dynamic_pointer_cast<OwnerRRef<py::object>>(rref_)->getValue();
+    // acquiring GIL as the return statement construct a new py::object from
+    // a const reference.
+    AutoGIL ag;
+    return value;
+  } else {
+    auto value =
+        std::dynamic_pointer_cast<OwnerRRef<IValue>>(rref_)->getValue();
+    // acquiring GIL as torch::jit::toPyObject creates new py::object without
+    // grabbing the GIL.
+    AutoGIL ag;
+    return torch::jit::toPyObject(std::move(value));
+  }
+}
+
+py::tuple PyRRef::pickle() const {
+  auto& ctx = RRefContext::getInstance();
+  auto rfd = ctx->prepareChildFork(rref_);
```

Comment:
Do you mean we only install the dispatch table when the RRef is going across RPC boundaries? What if the application pickles and then unpickles an RRef to get a new one, should we consider that as a new RRef instance and reference counting it?

Reply:
yes that's what I mean. Because those behaviors are specific to RPC protocol. 

For the pickle and unpickle example, it is a new rref instance and we should reference count it. But it's probably more natural to ref count it when the new rref is created, rather than increasing the ref count at pickle time, if the pickle happens on the same machine. 
 
I think it's fine to keep what we have, since doing dispatch table is over engineering at this point. It would be good to document is properly so there is no surprise.